### ISSUE TITLE
Where would you add ThreadLocal/MDC type data

### ISSUE THREAD

**Issue by chrisjowen** [tomtom-international/kotlin-tools#29]
Hi Guys, 

Sorry this isn't really a issue, just trying to figure my way through this and as part of tracing I need to get context specific to he executing thread in the same way as you might use MDC in logging frameworks. Is there anything I missing that allows this already?

Digging around, and having limited knowladge of corountines I suspect that at the point of calling a method on the proxy object the invocation of this would be where we might add ThreadLocal context info to the `TraceEvent`. Somewhere here https://github.com/tomtom-international/kotlin-tools/blob/master/traceevents/src/main/java/com/tomtom/kotlin/traceevents/Tracer.kt#L299 

If I understand correctly this must be the place as when the event consumer is `consumeTraceEvent` is called this is happening in its own thread processing events syncronously from the channel. 

Again my knowlage here is limited, but does this look like the place where I might set this kind of informaiton or is there another hook I do not see?

Perhaps its easier if I specify the use case, I want to emit these events in from a web request at which point we set a corrlation id, each event I emit I would want to pass through this ID within a context so when I am consuming them I can group them. I know I can do this manually from the calling point, just wondering if this is/could be done automatically somehow

**Comment by RijnBuve-TomTom** [comment]
Hi chrisjowen, if you want to add thread-specific information to the event, indeed the `Tracer.invoke()` call is the location where you could add that (line 299, creating an instance of `TraceEvent`).

The event will subsequently be put in a queue and read from the queue and executed on other threads, so you can't do it elsewhere.

Alternatively, and perhaps even better, you might wish to consider using the `context` parameter for your grouping instead. If you can create different `Tracer`s for different contexts, then you don't need to adapt the library at all. See, for example, `LogContextTest.kt` line 44, `tracer declared in instance`.

Let me know if this helps, just reply to this thread.

**Comment by chrisjowen** [comment]
Hi @RijnBuve-TomTom thanks for the quick response. I did think about this but it seemed a bit limited, please correct me if I am wrong but from what I saw:

- The context is just a single string and I would like to maybe attach several items to the context during the run of the thing I want to trace i.e. just a `Map<String, Any>`
- Secondly context seems to also be used for filtering when attaching consumers which I don't want to do, I need to capture all trace logs and store the context would be for filtering later on in a UI

I was also a little worried about what the correct lifecycle for these things are, in all examples (including the test) they are initiated as part of a singleton companion object so I expect them to life for the lifecycle of the application. 

I think with my example where I would need a correlationId per request this would mean creating a new Tracer per request. Then I am either passing this around somehow or managing the tracer itself on some ThreadLocal context or simmilar. 

Honestly I don't like those ideas and I worried how this would be disposed of correctly i.e. after the request has ended and all messages had been emitted. I don't understand the internals enough to know if this is viable. Does this make sense, and are my concerns valid in anyway?

**Comment by RijnBuve-TomTom** [comment]
Hi @chrisjowen,

Some thoughts from my side here.

- The `context` is indeed a string, and as such not great for passing other data. I can see why you don't want that if you want to attach multiple things. (JSON-ing your data into the context also doesn't sound like a good idea.)
- As for the semantics of `context`: it was added to be able differentiate between tracers of the class in different threads or processes; in that sense your use of `context` would still kind of be in line with the original idea.
- You are right about the lifecycle thing. Creating a tracer per request may not be what you want to do. I thought you'd create a tracer per thread, but your threads are probably not long-lived - I guess they handle 1 web request and die.

Let me think a bit about it... Meanwhile: would it be possible for you to "just add" the additional info as a parameter to every trace call? So, rather than just issuing an event like `login(userName)` you'd have something like `login(tracerContext(), userName)`. Refactoring that to the former call would be a breeze, once/if the library supports adding your own context parameters.

Regards,
-Rijn

PS. We publish the library on Maven Central. You can find the library [here](https://search.maven.org/search?q=g:com.tomtom.kotlin%20AND%20a:kotlin-tools).

**Comment by chrisjowen** [comment]
Yeah sure, I did hack away at this but with not much thought beyond my immediate uses. So I had no intention of raising any PRs or making sure my code was up to scratch for you guys or anything. 

I assume your version will be more thought out but heres what I did: https://github.com/tomtom-international/kotlin-tools/compare/master...shift-fwd:master

1. I pretty much just ripped off them implementation of slf4j's https://github.com/qos-ch/slf4j/blob/master/slf4j-api/src/main/java/org/slf4j/helpers/BasicMDCAdapter.java and added a few things then renamed it to `TracerDiagnosticContext` and

2. Updated the `TraceEvent` class to have another property `val eventDiagnosticContext: Map<String, Any?>? = emptyMap()`  

3. On invoking the proxy I added set it `eventDiagnosticContext = TraceContext.getCopyOfContextMap()` 

4. Updated the `createLogMessage` to serialise the context


This would limit this to all tracers sharing the same context (per thread) though.

From my view I was only interested `GenericTraceEventConsumer` whereby I get the single `consumeTraceEvent` method.

Not sure how you would make any additional context available for for the more specific `TraceEventConsumer` though as the interface signature will not have an argument for the context.
